recipes-app/iot2050-firmware-update/files/iot2050-firmware-update.tmpl

```diff
@@ -647,53 +683,73 @@ def main(argv):
     parser.add_argument('-n', '--no-backup',
                         help='Do not generate a firmware backup',
                         action='store_true')
+    parser.add_argument('-d', '--backup-dir',
+                        help='Specify the firmware backup directory',
+                        nargs=1,
+                        default=os.getenv('HOME'))
+    parser.add_argument('-q', '--quiet',
+                        help='Update firmware quietly, assume "yes" as answer to \
+                        all prompts except reboot is "no" and run non-interactively',
+                        action='store_true')
     try:
         args = parser.parse_args()
     except IOError as e:
         print(e.strerror, file=sys.stderr)
-        exit(1)
+        return ErrorCode.INVALID_ARG.value
 
-    update_input = input(
-        "\nWarning: Update may render device unbootable. Continue (y/N)? ")
+    update_input = the_input(args.quiet, 'y',
+        "\n" + BOLD + "Warning: Update may render device unbootable." + ENDC + " Continue (y/N)? ")
     if not update_input == "y":
-        sys.exit(1)
+        return ErrorCode.CANCELED.value
 
     if args.force or args.reset:
-        erase_env_input = input("\nWarning: All U-Boot environment variables will be reset to factory settings. Continue (y/N)? ")
+        erase_env_input = the_input(args.quiet, 'y',
+                "\n" + BOLD + "Warning: All U-Boot environment variables will be reset to factory settings." + ENDC + " Continue (y/N)? ")
         if not erase_env_input == "y":
-            sys.exit(1)
+            return ErrorCode.CANCELED.value
     try:
         if args.force:
             # When forced, ignore all the checking and call the underlying updater directly
-            updater = FirmwareUpdate(args.firmware)
+            updater = FirmwareUpdate(args.firmware, args.backup_dir)
         elif args.rollback:
-            updater = RollbackFirmware(args.firmware)
+            updater = RollbackFirmware(args.firmware, args.backup_dir)
         else:
-            updater = ManagedFirmwareUpdate(args.firmware)
+            updater = ManagedFirmwareUpdate(args.firmware, args.backup_dir)
 
         if args.rollback:
-            rollback_confirm = input("\nRollback the firmware to the version before the upgrade (y/N)?")
+            rollback_confirm = the_input(args.quiet, 'y',
+                    "\nRollback the firmware to the version before the upgrade (y/N)?")
             if not rollback_confirm == "y":
-                sys.exit(1)
-            updater.rollback_firmware()
-            sys.exit(1)
+                return ErrorCode.CANCELED.value
+            ret = updater.rollback_firmware()
```

Comment:
What if an UpgradeError is thrown during update_firmware method? The code will jump to line 753 and try the rollback again? What will be the return code at the end, INTERNAL_ERR?

Reply:
I think, maybe we should have two different exception. UpgradeError for normal process and UpgradeRollbackError for rollback process :)